auditor/src/routes/advanced_record_filters.rs

```diff
@@ -0,0 +1,193 @@
+                       GROUP BY s.record_id
+                       ) m ON m.record_id = a.record_id
+                   LEFT JOIN (
+                       WITH subquery AS (
+                          SELECT 
+                              c.id as cid,
+                              COALESCE(array_agg(row(s.name, s.value)::score) FILTER (WHERE s.name IS NOT NULL AND s.value IS NOT NULL), '{}'::score[]) as scores
+                          FROM components as c
+                          LEFT JOIN components_scores as cs
+                          ON c.id = cs.component_id
+                          LEFT JOIN scores as s
+                          ON cs.score_id = s.id
+                          GROUP BY c.id
+                       )
+                       SELECT rc.record_id as id, array_agg(row(c.name, c.amount, sq.scores)::component) as components
+                       FROM records_components AS rc
+                       LEFT JOIN components as c
+                       ON rc.component_id = c.id
+                       LEFT JOIN subquery AS sq
+                       ON sq.cid = rc.component_id
+                       GROUP BY rc.record_id
+                   ) css ON css.id = a.id
+               ");
+
+    if let Some(filters) = &filters {
+        if filters.start_time.is_some() || filters.stop_time.is_some() || filters.runtime.is_some()
+        {
+            query.push(" WHERE ".to_string());
+            if let Some(start_time_filters) = &filters.start_time {
+                if let Some(operators) = get_operator(start_time_filters) {
+                    for operator in operators {
+                        let formatted_datetime = operator.1.format("%Y-%m-%d %H:%M:%S").to_string();
+                        query.push(format!(
+                            "a.start_time {} '{}' and ",
+                            operator.0, &formatted_datetime
+                        ));
+                    }
+                }
+            } else {
+                println!("start_time is not specified to query");
+            }
+
+            if let Some(stop_time_filters) = &filters.stop_time {
+                if let Some(operators) = get_operator(stop_time_filters) {
+                    for operator in operators {
+                        let formatted_datetime = operator.1.format("%Y-%m-%d %H:%M:%S").to_string();
+                        query.push(format!(
+                            "a.stop_time {} '{}' and ",
+                            operator.0, &formatted_datetime
+                        ));
+                    }
+                }
+            } else {
+                println!("stop_time is not specified to query");
+            }
+
+            if let Some(runtime_filters) = &filters.runtime {
+                if let Some(operators) = get_operator(runtime_filters) {
+                    for operator in operators {
+                        query.push(format!("a.runtime {} {} and ", operator.0, operator.1));
```

Comment:
In the queries above we always a trailing `and`, which is ok, because either we append more conditions here or the `a.runtime IS NOT NULL` check from below In the case of `runtime_filters = None`. However, when `runtime_filters` are not `None`, don't we end up with a trailing `and` at the end of the query?